File: ios/Sources/SpeechOutput.swift
import AVFAudio
import Foundation

/// Spoken translation output, and the decisions behind it.
///
/// Nothing here can be asserted by listening to a speaker, so the same split the session-comfort
/// work uses applies again: every decision is a pure function or a small state machine over an
/// injected sink, and the `AVSpeechSynthesizer` call is the thin shell around it. The tests
/// exercise the decisions; the sound itself is verified on device.

// MARK: - Copy

enum SpeechOutputCopy {
  static var replayAction: String {
    String(localized: "Play translation", comment: "Accessibility label for a bubble's replay control")
  }
  /// Not "again": with speech on tap only, this control is the first play as often as it is the
  /// second, and a hint that promises a repeat is wrong on every bubble nobody has played yet.
  static var replayHint: String {
    String(localized: "Reads this translation aloud.",
           comment: "Accessibility hint for a bubble's play control")
  }
}

/// The one speech glyph left on the screen.
///
/// There used to be two, and the pair was the problem: a crossed-out speaker in the status strip
/// saying "this app is currently silent" over a live speaker on every bubble is one screen making
/// two contradictory claims about sound. With speech on tap only there is no state to draw, so
/// only the action remains, and it takes the platform's sign for an action that plays something.
enum SpeechGlyph {
  static let replay = "play.circle"
}

// MARK: - What gets spoken

extension SessionState {
  /// The states in which the microphone is open and the recognizer owns the audio session.
  /// Nothing is ever spoken in these; `translating` is not one of them, because the recognizer has
  /// already been stopped by the time a translation is under way.
  var isRecognizerLive: Bool {
    switch self {
    case .listening, .finalizing: true
    default: false
    }
  }
}

extension ConversationItem {
  /// The text this bubble can speak: a finished translation and nothing else. A bubble that is
  /// still recognizing, still translating, or whose translation failed has nothing to say, which
  /// is exactly when the play control is hidden.
  var speakableTranslation: String? {
    guard case .translated = state, let translation else { return nil }
    return translation.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : translation
  }
}

/// Whether a bubble speaks, and in which voice language.
///
/// One rule, and now one caller: a tap on that bubble's play control. Nothing speaks on delivery
/// any more, so this no longer arbitrates between an automatic announcement and a replay, and
/// there is no mute to consult: the tap is the consent.
enum SpokenTranslation: Equatable {
  case speak(text: String, languageCode: String)
  case silent

  static func decision(for item: ConversationItem) -> SpokenTranslation {
    guard let text = item.speakableTranslation else { return .silent }
    return .speak(text: text, languageCode: item.targetLanguage.code)
  }
}

// MARK: - Voice matching

/// Picking the voice a translation is read in. The reading language is a Hy-MT2 code such as `ko`
/// or `zh-Hant`, and the installed voices are concrete locales such as `ko-KR` or `zh-TW`, so the
/// two have to be matched rather than compared.
enum SpeechVoice {
  /// The best installed voice language for a target code: an exact match first, then the variant
  /// the code most likely implies (`zh-Hant` picks `zh-TW` over `zh-CN`), then any voice for the
  /// same language, and nil when the device has none at all. Nil means the translation stays
  /// silent rather than being read out in the wrong language.
  static func match(for code: String, in available: [String]) -> String? {
    let target = Locale.Language(identifier: code)
    guard let language = target.languageCode?.identifier else { return nil }
    if let exact = available.first(where: { $0.caseInsensitiveCompare(code) == .orderedSame }) {
      return exact
    }
    let candidates = available
      .filter { Locale.Language(identifier: $0).languageCode?.identifier == language }
      .sorted { $0.localizedStandardCompare($1) == .orderedAscending }
    guard candidates.count > 1 else { return candidates.first }
    let implied = subtags(of: target.maximalIdentifier).subtracting([language])
    return candidates.first { !subtags(of: $0).subtracting([language]).intersection(implied).isEmpty }
      ?? candidates.first
  }

  private static func subtags(of identifier: String) -> Set<String> {
    Set(
      identifier.replacingOccurrences(of: "_", with: "-")
        .split(separator: "-").map(String.init)
    )
  }
}

// MARK: - Audio session handoff

/// The audio session as the speech output needs it. `AVAudioSession` in the app, a spy in tests.
protocol SpeechAudioSession {
  func activatePlayback() throws
  func deactivate()
}

/// Speech recognition runs the session as `.record` with `.measurement` mode, which cannot play
/// anything, so speaking takes the session over as `.playback` and hands it straight back.
/// `.notifyOthersOnDeactivation` on the way out is what lets whatever was playing before resume,
/// and what leaves the session free for the next push-to-talk to claim as `.record` again.
struct SystemSpeechAudioSession: SpeechAudioSession {
  func activatePlayback() throws {
    let session = AVAudioSession.sharedInstance()
    try session.setCategory(.playback, mode: .spokenAudio, options: [.duckOthers])
    try session.setActive(true, options: [])
  }

  func deactivate() {
    try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
  }
}

/// Owns the record/playback handoff, and remembers which side currently holds the session.
///
/// The memory is the point. A translation that finishes while another is being spoken must not
/// tear the session down and rebuild it mid-sentence, and the delegate callback that says an
/// utterance ended arrives after the next push-to-talk may already have claimed the microphone.
/// Both are the same bug, and both are prevented by only ever writing a change: `endPlayback`
/// after the session was already handed back does nothing at all.
@MainActor
final class SpeechAudioCoordinator {
  private(set) var isPlaybackActive = false

  private let session: any SpeechAudioSession

  init(session: any SpeechAudioSession = SystemSpeechAudioSession()) {
    self.session = session
  }

  /// Claims the session for speech. Returns whether it is safe to speak: a session that refuses
  /// to activate stays unclaimed, so nothing is spoken into a route that is not there and the
  /// next attempt tries again from scratch.
  @discardableResult
  func beginPlayback() -> Bool {
    guard !isPlaybackActive else { return true }
    do {
      try session.activatePlayback()
      isPlaybackActive = true
    } catch {
      isPlaybackActive = false
    }
    return isPlaybackActive
  }

  /// Hands the session back, once. Called both when speech ends on its own and, synchronously,
  /// before a turn starts recording.
  func endPlayback() {
    guard isPlaybackActive else { return }
    isPlaybackActive = false
    session.deactivate()
  }
}

// MARK: - Speech output

/// Anything that can read a translation aloud. `AVSpeechSynthesizer` in the app, a fake in tests.
@MainActor
protocol SpeechOutput: AnyObject {
  /// Speaks `text` in the best available voice for `languageCode`, replacing anything already
  /// being spoken. A language with no installed voice speaks nothing.
  func speak(text: String, languageCode: String)
  /// Stops immediately and hands the audio session back, synchronously, so a push-to-talk that
  /// interrupts a sentence can claim the microphone on the very next line.
  func stop()
  var isSpeaking: Bool { get }
}

@MainActor
final class SystemSpeechOutput: NSObject, SpeechOutput, AVSpeechSynthesizerDelegate {
  private let synthesizer: AVSpeechSynthesizer
  private let audio: SpeechAudioCoordinator
  private let voiceLanguages: () -> [String]

  var isSpeaking: Bool { synthesizer.isSpeaking }

  init(synthesizer: AVSpeechSynthesizer = AVSpeechSynthesizer(),
       audio: SpeechAudioCoordinator? = nil,
       voiceLanguages: @escaping () -> [String] = {
         AVSpeechSynthesisVoice.speechVoices().map(\.language)
       }) {
    self.synthesizer = synthesizer
    self.audio = audio ?? SpeechAudioCoordinator()
    self.voiceLanguages = voiceLanguages
    super.init()
    synthesizer.delegate = self
  }

  func speak(text: String, languageCode: String) {
    guard let language = SpeechVoice.match(for: languageCode, in: voiceLanguages()),
          let voice = AVSpeechSynthesisVoice(language: language) else {
      // No voice for this language at all: say nothing rather than read it in another accent.
      stop()
      return
    }
    // The newest translation wins. The session is deliberately not released here, so replacing a
    // sentence is a cut rather than a route change the speaker can hear.
    if synthesizer.isSpeaking { synthesizer.stopSpeaking(at: .immediate) }
    guard audio.beginPlayback() else { return }
    let utterance = AVSpeechUtterance(string: text)
    utterance.voice = voice
    synthesizer.speak(utterance)
  }

  func stop() {
    if synthesizer.isSpeaking { synthesizer.stopSpeaking(at: .immediate) }
    audio.endPlayback()
  }

  nonisolated func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer,
                                     didFinish utterance: AVSpeechUtterance) {
    Task { @MainActor [weak self] in self?.utteranceEnded() }
  }

  nonisolated func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer,
                                     didCancel utterance: AVSpeechUtterance) {
    Task { @MainActor [weak self] in self?.utteranceEnded() }
  }

  /// The end of one utterance is only the end of speech when nothing took its place, and never a
  /// reason to touch a session the recognizer has since claimed.
  private func utteranceEnded() {
    guard !synthesizer.isSpeaking else { return }
    audio.endPlayback()
  }
}

File: android/app/src/main/java/ai/zetic/realtimetranslate/SpeechOutput.kt
package ai.zetic.realtimetranslate

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioFocusRequest
import android.media.AudioManager
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import java.util.Locale

/**
 * Spoken translation output, and the decisions behind it.
 *
 * Completed translations are spoken only when the user taps replay. The voice is the platform's
 * own synthesizer; no model is downloaded and nothing is sent anywhere.
 *
 * Nothing here can be asserted by listening to a speaker, so the split the session-comfort work
 * uses applies again: every decision is a pure function or a small state machine over an injected
 * seam, and the `TextToSpeech` call is the thin shell around it. The tests exercise the decisions;
 * the sound itself is verified on a device.
 */

// region What gets spoken

/**
 * The states in which the microphone is open and the recognizer owns the audio route. Nothing is
 * ever spoken in these. `Translating` is not one of them: the recognizer has already been released
 * by the time a translation is under way.
 */
val SessionUiState.isRecognizerLive: Boolean
    get() = phase in setOf(
        SessionPhase.ListeningA,
        SessionPhase.ListeningB,
        SessionPhase.FinalizingA,
        SessionPhase.FinalizingB,
    )

/**
 * The text this bubble can speak: a finished translation and nothing else. A bubble that is still
 * recognizing, still translating, or whose translation failed has nothing to say, which is exactly
 * when the replay control is absent rather than disabled.
 */
val ConversationItem.speakableTranslation: String?
    get() = translation?.takeIf { translationError == null && it.isNotBlank() }

/**
 * Whether a bubble can be replayed, and in which language. Nothing is spoken automatically.
 */
sealed interface SpokenTranslation {
    data class Speak(val text: String, val languageCode: String) : SpokenTranslation
    data object Silent : SpokenTranslation

    companion object {
        fun decision(item: ConversationItem): SpokenTranslation {
            val text = item.speakableTranslation ?: return Silent
            return Speak(text, item.targetLanguage.code)
        }
    }
}

/**
 * The replay control on a translated bubble. Absent when there is nothing to play, and present but
 * disabled while the recognizer holds the microphone: a control that answers a tap
 * with silence reads as broken rather than busy.
 */
object ReplayControl {
    fun isPresent(item: ConversationItem): Boolean = item.speakableTranslation != null

    fun isEnabled(item: ConversationItem, isRecognizerLive: Boolean): Boolean =
        isPresent(item) && !isRecognizerLive
}

// endregion

// region Voice matching

/**
 * Picking the voice a translation is read in. The reading language is a Hy-MT2 code such as `ko` or
 * `zh-Hant`; the installed voices are concrete locales such as `ko-KR` or `zh-TW`. The two are
 * matched rather than compared, in three steps: an exact match for the code, then the variant that
 * code implies, then any installed voice for the same language.
 *
 * No match means the translation stays silent. A language with no installed voice is never read out
 * in another language's voice.
 */
object SpeechVoiceMatching {
    fun match(code: String, available: List<Locale>): Locale? {
        val target = Locale.forLanguageTag(code)
        val language = target.language.takeIf { it.isNotEmpty() } ?: return null
        available.firstOrNull { it.toLanguageTag().equals(code, ignoreCase = true) }?.let { return it }
        val candidates = available
            .filter { it.language.equals(language, ignoreCase = true) }
            .sortedBy { it.toLanguageTag() }
        if (candidates.size <= 1) return candidates.firstOrNull()
        val implied = subtags(code) + likelyVariants.getOrElse(code) { emptySet() }
        return candidates.firstOrNull { subtags(it.toLanguageTag()).any(implied::contains) } ?: candidates.first()
    }

    /** Everything after the primary subtag, upper-cased so `Hant` and `HANT` compare equal. */
    private fun subtags(tag: String): Set<String> =
        tag.split('-', '_').drop(1).filter(String::isNotEmpty).map { it.uppercase(Locale.ROOT) }.toSet()

    /**
     * The variant a bare reading code implies, for the Hy-MT2 languages a phone plausibly carries
     * more than one voice for. The same table the recognizer's language matching uses, because the
     * question is the same one.
     */
    private val likelyVariants = mapOf(
        "ar" to setOf("SA"),
        "bn" to setOf("BD"),
        "de" to setOf("DE"),
        "en" to setOf("US"),
        "es" to setOf("ES"),
        "fr" to setOf("FR"),
        "it" to setOf("IT"),
        "ms" to setOf("MY"),
        "nl" to setOf("NL"),
        "pt" to setOf("BR"),
        "ta" to setOf("IN"),
        "ur" to setOf("PK"),
        "zh" to setOf("CN", "HANS"),
        "zh-Hant" to setOf("TW", "HANT"),
    )
}

// endregion

// region Audio focus

/** The audio focus the speech output needs. `AudioManager` in the app, a spy in tests. */
interface SpeechAudioFocus {
    /** @return whether focus was granted. A refusal means nothing is spoken. */
    fun request(onInterrupted: () -> Unit): Boolean
    fun abandon()
}

/**
 * Owns the focus handoff, and remembers which side currently holds it.
 *
 * The memory is the point. A translation that finishes while another is being spoken must not drop
 * and re-take focus mid-sentence, and the utterance-done callback that says speech ended arrives
 * after the next push-to-talk may already have claimed the microphone. Both are the same bug, and
 * both are prevented by only ever writing a change: [release] after focus was already handed back
 * does nothing at all.
 */
class SpeechAudioCoordinator(private val focus: SpeechAudioFocus) {
    var isHoldingFocus: Boolean = false
        private set

    /**
     * Claims focus for speech. Returns whether it is safe to speak: a refusal leaves focus
     * unclaimed, so nothing is spoken over whatever holds the route, and the next translation
     * tries again from scratch.
     */
    fun claim(onInterrupted: () -> Unit = {}): Boolean {
        if (isHoldingFocus) return true
        isHoldingFocus = focus.request(onInterrupted)
        return isHoldingFocus
    }

    /** Hands focus back, once. Called when speech ends and, synchronously, before a turn starts. */
    fun release() {
        if (!isHoldingFocus) return
        isHoldingFocus = false
        focus.abandon()
    }
}

/**
 * Transient focus that ducks rather than pauses: a translation is one short sentence, and whatever
 * else the phone is playing should dip under it and come straight back.
 */
class AndroidSpeechAudioFocus(context: Context) : SpeechAudioFocus {
    private val applicationContext = context.applicationContext
    private val manager = applicationContext.getSystemService(AudioManager::class.java)
    private val routeLoss = AudioRouteLossMonitor(applicationContext)
    private var request: AudioFocusRequest? = null
    private var interruption: (() -> Unit)? = null

    private val listener = AudioManager.OnAudioFocusChangeListener { change ->
        if (SpeechPlaybackFocusContract.interrupts(change)) interruption?.invoke()
    }

    override fun request(onInterrupted: () -> Unit): Boolean {
        val audioManager = manager ?: return false
        interruption = onInterrupted
        val focusRequest = AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK)
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_ASSISTANT)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                    .build(),
            )
            .setOnAudioFocusChangeListener(listener)
            .build()
        request = focusRequest
        val granted = audioManager.requestAudioFocus(focusRequest) == AudioManager.AUDIOFOCUS_REQUEST_GRANTED
        if (!granted) {
            request = null
            interruption = null
        } else {
            routeLoss.start { interruption?.invoke() }
        }
        return granted
    }

    override fun abandon() {
        interruption = null
        routeLoss.stop()
        val audioManager = manager ?: return
        request?.let { audioManager.abandonAudioFocusRequest(it) }
        request = null
    }
}

internal object SpeechPlaybackFocusContract {
    fun interrupts(change: Int): Boolean = change in setOf(
        AudioManager.AUDIOFOCUS_LOSS,
        AudioManager.AUDIOFOCUS_LOSS_TRANSIENT,
        AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK,
    )

    fun interruptsRouteChange(action: String?): Boolean = AudioRouteLossContract.interrupts(action)
}

// endregion

// region Speech output

/** Anything that can read a translation aloud. `TextToSpeech` in the app, a fake in tests. */
interface SpeechOutput {
    /**
     * Speaks [text] in the best available voice for [languageCode], replacing anything already
     * being spoken. A language with no installed voice speaks nothing rather than reading the
     * sentence in the wrong accent.
     */
    fun speak(text: String, languageCode: String)

    /** Stops immediately and hands audio focus back, so an interrupting turn can start clean. */
    fun stop()

    fun shutdown()
}

/**
 * The platform synthesizer, behind [SpeechOutput].
 *
 * The engine is asynchronous to start and the first translation can land before it is ready, so
 * every call guards on the init callback rather than assuming it. Focus is claimed once and held
 * across a replacement, because replacing one sentence with a newer one is a cut, not a route
 * change the listener should hear.
 */
class AndroidSpeechOutput(
    context: Context,
    private val audio: SpeechAudioCoordinator = SpeechAudioCoordinator(AndroidSpeechAudioFocus(context)),
) : SpeechOutput {
    private var engine: TextToSpeech? = null
    private var isReady = false
    private var utteranceCount = 0

    init {
        engine = TextToSpeech(context.applicationContext) { status ->
            isReady = status == TextToSpeech.SUCCESS
        }.apply {
            setOnUtteranceProgressListener(object : UtteranceProgressListener() {
                override fun onStart(utteranceId: String?) = Unit
                override fun onDone(utteranceId: String?) = releaseIfIdle()
                override fun onStop(utteranceId: String?, interrupted: Boolean) = releaseIfIdle()

                @Deprecated("Replaced by onError(String, Int)", ReplaceWith(""))
                override fun onError(utteranceId: String?) = releaseIfIdle()
                override fun onError(utteranceId: String?, errorCode: Int) = releaseIfIdle()
            })
        }
    }

    private val availableLocales: List<Locale>
        get() = runCatching { engine?.availableLanguages?.toList() }.getOrNull().orEmpty()

    override fun speak(text: String, languageCode: String) {
        val tts = engine?.takeIf { isReady } ?: return
        val voice = SpeechVoiceMatching.match(languageCode, availableLocales)
        if (voice == null) {
            // No voice for this language at all: say nothing rather than read it in another accent.
            stop()
            return
        }
        val setLanguage = runCatching { tts.setLanguage(voice) }.getOrDefault(TextToSpeech.LANG_MISSING_DATA)
        if (setLanguage == TextToSpeech.LANG_MISSING_DATA || setLanguage == TextToSpeech.LANG_NOT_SUPPORTED) {
            stop()
            return
        }
        // Newest wins. Focus is deliberately not released here: the cut should not be audible.
        if (!audio.claim(::stop)) return
        utteranceCount += 1
        tts.speak(text, TextToSpeech.QUEUE_FLUSH, null, "translation-$utteranceCount")
    }

    override fun stop() {
        runCatching { engine?.stop() }
        audio.release()
    }

    override fun shutdown() {
        stop()
        runCatching { engine?.shutdown() }
        engine = null
        isReady = false
    }

    /** The end of one utterance is only the end of speech when nothing took its place. */
    private fun releaseIfIdle() {
        if (engine?.isSpeaking == true) return
        audio.release()
    }
}

// endregion
